lib/Dialect/FIRRTL/FIRRTLFolds.cpp

```diff
@@ -124,10 +130,72 @@ OpFoldResult ConstantOp::fold(ArrayRef<Attribute> operands) {
   return valueAttr();
 }
 
+/// Check if the results of \p op are of integer type and with known bitwidth.
+/// Can be used to determine if any fold is legal.
+bool hasKnownWidths(Operation *op) {
+  for (auto resTy : op->getResultTypes()) {
+    auto ty = resTy.cast<FIRRTLType>();
+    if (!ty.isa<IntType>() || ty.getBitWidthOrSentinel() == -1)
+      return false;
+  }
+  return true;
+}
+
+// Width extension of a constant int depends only on the sign of the operand.
+APInt extendConstantToWidth(APInt old, bool isSigned, unsigned newWidth) {
+  if (isSigned)
+    return old.sext(newWidth);
+  return old.zext(newWidth);
+}
+
 //===----------------------------------------------------------------------===//
 // Binary Operators
 //===----------------------------------------------------------------------===//
 
+OpFoldResult AddPrimOp::fold(ArrayRef<Attribute> operands) {
+  /// Any folding here requires a bitwidth extension.
+
+  /// If both operands are constant, and the result is integer with known
+  /// widths, then perform constant folding. Cannot use constFoldBinaryOp, since
+  /// the width of the constant is different from operands.
+  return constFoldFIRRTLBinaryOp<APInt>(
+      operands, getOperand(0).getType().cast<IntType>().isUnsigned(),
```

Comment:
I think that each of these can be slightly simplified to use `getType().cast<IntType>().isUnsigned()`. instead of `getOperand(0).getType()...`, since the sign of the result type is always the sign of the operand type.